### PR TITLE
Iterated impes 2

### PR DESCRIPTION
**Pull request by anne-glerum**
Rebuild all preconditioners every nonlinear iteration

### CODE REVIEW

File: source/simulator/core.cc [hunk 1]
```diff
@@ -1673,9 +1676,10 @@ namespace aspect
                 rebuild_stokes_matrix = rebuild_stokes_preconditioner = true;
 
               assemble_stokes_system();
+              build_stokes_preconditioner();
+
               if (iteration == 0)
                 {
-                  build_stokes_preconditioner();
                   initial_stokes_residual = compute_initial_stokes_residual();
                 }
```
Comment (jdannberg) [inline source/simulator/core.cc]: You don't need the brackets {} here.

File: source/simulator/core.cc [hunk 2]
```diff
@@ -1703,7 +1707,9 @@ namespace aspect
                 max = std::max(stokes_residual/initial_stokes_residual, max);
               if (initial_temperature_residual>0)
                 max = std::max(temperature_residual/initial_temperature_residual, max);
-              pcout << "      residual: " << max << std::endl;
+              pcout << "      Residual: " << max << std::endl;
```
Comment (jdannberg) [inline source/simulator/core.cc]: Can you rename this to "Total nonlinear residual"?

File: source/simulator/core.cc [hunk 3]
```diff
@@ -1651,8 +1651,11 @@ namespace aspect
                   assemble_advection_system (AdvectionField::composition(c));
                   build_advection_preconditioner(AdvectionField::composition(c),
                                                  C_preconditioner);
+
                   if (iteration == 0)
-                    initial_composition_residual[c] = system_rhs.block(introspection.block_indices.compositional_fields[c]).l2_norm();
+                    {
+                      initial_composition_residual[c] = system_rhs.block(introspection.block_indices.compositional_fields[c]).l2_norm();
+                    }
```
Comment (jdannberg) [inline source/simulator/core.cc]: Same here, you don't need the brackets { }.

**Comment by jdannberg** [comment]
Can you update the test outputs?

The following tests FAILED:
	 42 - composition_reaction_iterated_IMPES (Failed)
	 82 - iterated_IMPES (Failed)
	 83 - iterated_IMPES_direct_solver (Failed)
	 84 - iterated_IMPES_residual (Failed)

**Comment by jdannberg** [comment]
Great work otherwise, thanks for doing this!

**Comment by jdannberg** [comment]
Great, thank you!
@tjhei Can you please update the tests and then merge?

**Comment by tjhei** [comment]
see #499

**Pull request merged**